README.md
**The modern [local Expo CLI](https://blog.expo.dev/the-new-expo-cli-f4250d8e3421) is included with the `expo` package and does not need to be installed separately. Run it with `npx expo`. Its source code lives in the [`expo/expo`](https://github.com/expo/expo/blob/main/packages/%40expo/cli) repo.**

---

# Assorted Expo Dev Tools

This repo consists of packages used in the legacy global Expo CLI and related tooling.

- [🗺 Project Layout](#-project-layout)
- [👏 Contributing](#-contributing)
- [❓ FAQ](#-faq)
- [💙 The Team](#-the-team)
- [License](#license)

## 📚 Documentation

<p>Learn about building and deploying universal apps <a aria-label="expo documentation" href="https://docs.expo.dev">in our official docs!</a></p>

- [Using Expo CLI](https://docs.expo.dev/workflow/expo-cli/)
- [App.json Configuration](https://docs.expo.dev/workflow/configuration/)
- [Building and Deploying apps](https://docs.expo.dev/introduction/walkthrough/#building-and-deploying)

## 🗺 Project Layout

<!-- Begin auto-generation -->

| Package                                                     | Version                                                                                                                                               |
| ----------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
| [**`expo-cli`**](./packages/expo-cli)                       | [![badges](https://img.shields.io/npm/v/expo-cli?color=32cd32&style=flat-square)](https://www.npmjs.com/package/expo-cli)                             |
| [**`expo-codemod`**](./packages/expo-codemod)               | [![badges](https://img.shields.io/npm/v/expo-codemod?color=32cd32&style=flat-square)](https://www.npmjs.com/package/expo-codemod)                     |
| [**`@expo/image-utils`**](./packages/image-utils)           | [![badges](https://img.shields.io/npm/v/@expo/image-utils?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/image-utils)           |
| [**`@expo/json-file`**](./packages/json-file)               | [![badges](https://img.shields.io/npm/v/@expo/json-file?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/json-file)               |
| [**`@expo/next-adapter`**](./packages/next-adapter)         | [![badges](https://img.shields.io/npm/v/@expo/next-adapter?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/next-adapter)         |
| [**`@expo/osascript`**](./packages/osascript)               | [![badges](https://img.shields.io/npm/v/@expo/osascript?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/osascript)               |
| [**`@expo/pkcs12`**](./packages/pkcs12)                     | [![badges](https://img.shields.io/npm/v/@expo/pkcs12?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/pkcs12)                     |
| [**`@expo/plist`**](./packages/plist)                       | [![badges](https://img.shields.io/npm/v/@expo/plist?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/plist)                       |
| [**`pod-install`**](./packages/pod-install)                 | [![badges](https://img.shields.io/npm/v/pod-install?color=32cd32&style=flat-square)](https://www.npmjs.com/package/pod-install)                       |
| [**`expo-pwa`**](./packages/pwa)                            | [![badges](https://img.shields.io/npm/v/expo-pwa?color=32cd32&style=flat-square)](https://www.npmjs.com/package/expo-pwa)                             |
| [**`@expo/schemer`**](./packages/schemer)                   | [![badges](https://img.shields.io/npm/v/@expo/schemer?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/schemer)                   |
| [**`uri-scheme`**](./packages/uri-scheme)                   | [![badges](https://img.shields.io/npm/v/uri-scheme?color=32cd32&style=flat-square)](https://www.npmjs.com/package/uri-scheme)                         |
| [**`@expo/webpack-config`**](./packages/webpack-config)     | [![badges](https://img.shields.io/npm/v/@expo/webpack-config?color=32cd32&style=flat-square)](https://www.npmjs.com/package/@expo/webpack-config)     |

## 👏 Contributing

If you like the Expo CLI and want to help make it better then check out our [contributing guide](/CONTRIBUTING.md)! Also check out the [Expo repo](http://github.com/expo/expo) to work on the Expo docs, modules, and components in the Expo SDK.

## ❓ FAQ

If you have questions about Expo and want answers, then check out our [Frequently Asked Questions](https://docs.expo.dev/introduction/faq/)!

If you still have questions you can ask them on our [forums](https://forums.expo.dev) or on Twitter [@Expo](https://twitter.com/expo).

## 💙 The Team

Curious about who makes Expo? Here are our [team members](https://expo.dev/about)!

## License

The Expo source code is made available under the [MIT license](LICENSE). Some of the dependencies are licensed differently, with the BSD license, for example.


CHANGELOG.md
# Changelog

This is the log of notable changes to Expo CLI and related packages.

## main

### 🛠 Breaking changes

### 🎉 New features

### 🧹 Chores

## [Sun, 10 Sep 2023 17:56:11 -0700](https://github.com/expo/expo-cli/commit/24eb3415f8e120e82a4f6dc8577d07fbe42ef854)

### 🛠 Breaking changes

### 🎉 New features

- [doctor] @expo/metro-config deep dependency check ([#4742](https://github.com/expo/expo-cli/issues/4742))
- [pod-install] show alternative message in managed projects ([#4566](https://github.com/expo/expo-cli/issues/4566))
- [create-expo] Bump @expo/package-manager for Bun support
- [create-expo] detect bun package manager ([#4752](https://github.com/expo/expo-cli/issues/4752))
- [webpack]: Bump expo to SDK 49 ([#4747](https://github.com/expo/expo-cli/issues/4747))
- [schemer]: additional validation for unsupported image formats ([#4764](https://github.com/expo/expo-cli/pull/4764))

### 🧹 Chores

### 🐛 Bug fixes

- [create-expo]: allow scoped template package names ([#4750](https://github.com/expo/expo-cli/issues/4750))

### 📦 Packages updated

- create-expo@2.1.1
- create-expo-app@2.1.1
- @expo/dev-tools@0.13.187
- expo-cli@6.3.12
- expo-doctor@1.1.3
- install-expo-modules@0.6.3
- pod-install@0.1.39
- expo-pwa@0.0.127
- @expo/webpack-config@18.1.3
- xdl@60.0.12

## [Wed Jul 5 17:04:00 2023 -0700](https://github.com/expo/expo-cli/commit/b417540058c85cc7c2b28c37aa2e2ce52b06b03b)

### 🧹 Chores

- [install-expo-modules] Update text re: Expo CLI installation

### 📦 Packages updated

- install-expo-modules@0.6.1

## [Wed, 5 Jul 2023 12:03:05 -0700](https://github.com/expo/expo-cli/commit/2bc53712a2d0c856b5504a77f402a08e3bef993f)

### 🎉 New features

- [install-expo-modules] add cli integration support ([#4728](https://github.com/expo/expo-cli/issues/4728))

### 🧹 Chores

- [cli] Improve the Expo CLI deprecation message
- [doctor] clean up check names ([#4727](https://github.com/expo/expo-cli/issues/4727))
- [github] Remove webpack workflows
- [pwa] drop custom config path

### 📦 Packages updated

- @expo/babel-preset-cli@0.3.1
- create-expo@2.0.3
- @expo/dev-tools@0.13.185
- expo-cli@6.3.9
- expo-doctor@1.1.1
- install-expo-modules@0.6.0
- expo-pwa@0.0.126
- @expo/webpack-config@18.1.1
- xdl@60.0.10

## [Tue, 30 May 2023 13:01:03 -0500](https://github.com/expo/expo-cli/commit/23cc30b081aabefc2b0d8f7139c001fd286322d8)

### 🛠 Breaking changes

### 🎉 New features

- [doctor] Move dependencies to devDependencies ([#4709](https://github.com/expo/expo-cli/issues/4709))
- Make `@expo/webpack-config` depend on `expo` ([#4703](https://github.com/expo/expo-cli/issues/4703))

### 🧹 Chores

### 🐛 Bug fixes

- [doctor] remove sh command from dep check for Windows compat ([#4706](https://github.com/expo/expo-cli/issues/4706))

### 📦 Packages updated

- @expo/dev-tools@0.13.184
- expo-cli@6.3.8
- expo-doctor@1.1.0
- expo-pwa@0.0.125
- @expo/webpack-config@18.1.0
- xdl@60.0.9

## [Thu, 27 Apr 2023 13:46:02 -0700](https://github.com/expo/expo-cli/commit/80d7d352b476aba692226367a3bb6a8e78bf6294)

### 🧹 Chores

- [cli] cross deploy to create-expo ([#4698](https://github.com/expo/expo-cli/issues/4698))
- [cli] Add link to new Expo CLI blog post
- [cli] add deprecated commands to helpgroup ([#4696](https://github.com/expo/expo-cli/issues/4696))
- [cli] replace deprecated commands with superceded ones ([#4693](https://github.com/expo/expo-cli/issues/4693))
- [cli] advise to use new doctor on SDK 46+ ([#4689](https://github.com/expo/expo-cli/issues/4689))
- [cli] deprecate publish:\* ([#4685](https://github.com/expo/expo-cli/issues/4685))

### 🐛 Bug fixes

- [doctor] specify schemer version ([#4687](https://github.com/expo/expo-cli/issues/4687))

### 📦 Packages updated

- create-expo-app@1.3.3
- expo-cli@6.3.7
- expo-doctor@1.0.6

## [Tue, 14 Feb 2023 18:55:17 -0800](https://github.com/expo/expo-cli/commit/043302f4d0fdd3f13a4bfc928bd45c8c5f501c3a)

### 🎉 New features

- [create-expo-app] add `--example` flag to initialize from example ([#4644](https://github.com/expo/expo-cli/issues/4644))
- [install-expo-modules] add sdk 48 support ([#4652](https://github.com/expo/expo-cli/issues/4652))

### 🧹 Chores

- [xdl] remove unused import from `createBundlesAsync` ([#4645](https://github.com/expo/expo-cli/issues/4645))

### 📦 Packages updated

- create-expo-app@1.3.2
- @expo/dev-tools@0.13.178
- expo-cli@6.3.1
- install-expo-modules@0.5.0
- xdl@60.0.3

## [Mon, 6 Feb 2023 20:37:17 -0800](https://github.com/expo/expo-cli/commit/36dbc671d41270885a046fd5848ddd3b039139ed)

### 🎉 New features

- [expo-cli] Bump @expo/dev-server to support the latest React Native release with Hermes sourcemap changes

### 🐛 Bug fixes

- [expo-cli] Bump @expo/apple-utils
- [next-adapter] drop next-expo script as it causes npm to fail ([#4637](https://github.com/expo/expo-cli/issues/4637))

### 📦 Packages updated

- create-expo-app@1.2.1
- @expo/dev-tools@0.13.177
- expo-cli@6.3.0
- @expo/next-adapter@5.0.2
- xdl@60.0.2

## [Fri, 27 Jan 2023 11:42:59 -0600](https://github.com/expo/expo-cli/commit/226ac07bd39ec14a97441365af0d2030a8ff0f59)

### 🐛 Bug fixes

- fix(webpack): add webpack-dev-server as a dependency of webpack-config ([#4636](https://github.com/expo/expo-cli/issues/4636))

### 📦 Packages updated

- @expo/dev-tools@0.13.176
- expo-cli@6.2.1
- @expo/webpack-config@18.0.1
- xdl@60.0.1

## [Fri, 27 Jan 2023 11:01:19 -0600](https://github.com/expo/expo-cli/commit/4e7dbf3adc90106a3d013cb5289c1b0e1ca06172)

### 🛠 Breaking changes

- [webpack] Upgrade to Webpack 5 ([#3763](https://github.com/expo/expo-cli/issues/3763))
- Remove Node.js limitations on the legacy CLI ([#4635](https://github.com/expo/expo-cli/issues/4635))

### 🎉 New features

- [webpack] Upgrade to Webpack 5 ([#3763](https://github.com/expo/expo-cli/issues/3763))

### 🧹 Chores

- [readme] Highlight the modern local CLI over everything else ([#4632](https://github.com/expo/expo-cli/issues/4632))

### 🐛 Bug fixes


common/data.js
export const loggers = [
  {
    id: 1,
    type: 'process',
    name: 'Trace Viewer',
    logs: [
      {
        timestamp: '12:10:22 PM',
        text: 'Dumping profile information...',
      },
      {
        timestamp: '12:10:22 PM',
        text: 'Your profile was saved at:',
      },
      {
        timestamp: '12:10:22 PM',
        text: '/tmp/dump_1521281422867.json',
      },
      {
        timestamp: '12:10:22 PM',
        text:
          'On Google Chrome navigate to chrome://tracing and then click on "load" to load and visualise your profile.',
      },
      {
        timestamp: '12:10:22 PM',
        text: 'This message is also printed to your console by the packager so you can copy it :)',
      },
    ],
  },
  {
    id: 2,
    type: 'process',
    name: 'Metro Bundler',
    logs: [
      {
        status: 'success',
        timestamp: '3:14:30 AM',
        text: 'Project opened! You can now use the preview button to view your project',
      },
      {
        timestamp: '3:51:29 AM',
        text: 'Publishing...',
      },
      {
        timestamp: '3:52:29 AM',
        text: 'Building iOS bundle',
      },
      {
        timestamp: '3:53:29 AM',
        text: 'Building JavaScript bundle: finished in 45901ms',
      },
      {
        timestamp: '3:54:29 AM',
        text: 'Building Android bundle',
      },
      {
        timestamp: '3:55:29 AM',
        text: 'Building JavaScript bundle: finished in 20030ms',
      },
      {
        timestamp: '3:56:29 AM',
        text: 'Uploading assets',
      },
      {
        timestamp: '3:57:29 AM',
        text: 'No assets changed, skipped',
      },
      {
        timestamp: '3:58:29 AM',
        text: 'Uploading JavaScript bundles',
      },
      {
        status: 'success',
        timestamp: '3:59:29 AM',
        text: 'Published to https://exp.host/@jimmylee/ice',
      },
    ],
  },
  {
    id: 3,
    type: 'device',
    name: 'Mordor',
    logs: [
      {
        timestamp: '3:14:46 AM',
        text: 'Streaming logs from Mordor...',
      },
      {
        timestamp: '3:14:48 AM',
        text: `

 _|_|_|_|
 _|        _|    _|  _|_|_|      _|_|
 _|_|_|      _|_|    _|    _|  _|    _|
 _|        _|    _|  _|    _|  _|    _|
 _|_|_|_|  _|    _|  _|_|_|      _|_|
                     _|
                     _|
      `,
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
    ],
  },
  {
    id: 4,
    type: 'device',
    name: 'Mirkwood',
    logs: [
      {
        timestamp: '3:14:46 AM',
        text: 'Streaming logs from Mirkwood...',
      },
      {
        timestamp: '3:14:48 AM',
        text: `

 _|_|_|_|
 _|        _|    _|  _|_|_|      _|_|
 _|_|_|      _|_|    _|    _|  _|    _|
 _|        _|    _|  _|    _|  _|    _|
 _|_|_|_|  _|    _|  _|_|_|      _|_|
                     _|
                     _|
      `,
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
    ],
  },
  {
    id: 5,
    type: 'device',
    name: 'Isengard',
    logs: [
      {
        timestamp: '3:14:46 AM',
        text: 'Streaming logs from Isengard...',
      },
      {
        timestamp: '3:14:48 AM',
        text: `

 _|_|_|_|
 _|        _|    _|  _|_|_|      _|_|
 _|_|_|      _|_|    _|    _|  _|    _|
 _|        _|    _|  _|    _|  _|    _|
 _|_|_|_|  _|    _|  _|_|_|      _|_|
                     _|
                     _|
      `,
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
      {
        status: 'warning',
        timestamp: '3:14:48 AM',
        text: 'THREE.WebGLRenderer: OES_texture_float_linear extension not supported.',
      },
    ],
  },
];

export const icons = [
  {
    id: 1,
    src: '/static/default-project-icons/bordered-1.png',
  },
  {
    id: 2,
    src: '/static/default-project-icons/bordered-2.png',
  },
  {
    id: 3,
    src: '/static/default-project-icons/bordered-3.png',
  },
  {
    id: 4,
    src: '/static/default-project-icons/bordered-4.png',
  },
  {
    id: 5,
    src: '/static/default-project-icons/bordered-5.png',
  },
  {
    id: 6,
    src: '/static/default-project-icons/bordered-6.png',
  },
  {
    id: 7,
    src: '/static/default-project-icons/bordered-7.png',
  },
  {
    id: 8,
    src: '/static/default-project-icons/bordered-8.png',
  },
  {
    id: 9,
    src: '/static/default-project-icons/bordered-9.png',
  },
  {
    id: 10,
    src: '/static/default-project-icons/bordered-10.png',
  },
  {
    id: 11,
    src: '/static/default-project-icons/bordered-11.png',
  },
  {
    id: 12,
    src: '/static/default-project-icons/bordered-12.png',
  },
  {
    id: 13,
    src: '/static/default-project-icons/bordered-13.png',
  },
  {
    id: 14,
    src: '/static/default-project-icons/bordered-14.png',
  },
  {
    id: 15,
    src: '/static/default-project-icons/bordered-15.png',
  },
];

const delay = async time =>
  await new Promise((resolve, reject) => {
    window.setTimeout(() => resolve(), time);
  });

function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

export const generateServerSideMockData = store => {
  store.dispatch({
    type: 'CONNECT_DEVICE',
    device: {
      id: 1,
      type: 'process',
      name: 'Metro Bundler',
      logs: [],
    },
  });
};

export const generateClientSideMockData = async props => {
  props.dispatch({
    type: 'CONNECT_DEVICE',
    device: {
      id: 2,
      type: 'device',
      name: 'Mordor',
      logs: [],
    },
  });

  await delay(2000);

  props.dispatch({
    type: 'CONNECT_DEVICE',
    device: {
      id: 3,
      type: 'device',
      name: 'Isengard',
      logs: [],
    },
  });

  /*
  await delay(2000);

  props.dispatch({
    type: 'CONNECT_DEVICE',
    device: {
      id: 4,
      type: 'device',
      name: 'Moria',
      logs: [],
    },
  });
  */

  await delay(2000);

  // NOTE(jim): Rapid fire 40 logs over 4 devices and tests if it scrolls to bottom.
  for (let i = 0; i < 40; i++) {
    props.dispatch({
      type: 'PUSH_LOG_TO_DEVICE',
      log: {
        deviceIndex: getRandomInt(0, 3),
        status: 'success',
        timestamp: new Date().toLocaleTimeString('en-US'),
        text:
          'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAH Hi I am flooding you with logs.',
      },
    });
  }

  await delay(400);

  // NOTE(jim): Rapid fire 40 logs over 4 devices and tests if it scrolls to bottom.
  for (let i = 0; i < 40; i++) {
    props.dispatch({
      type: 'PUSH_LOG_TO_DEVICE',
      log: {
        deviceId: getRandomInt(1, 3),
        status: 'success',
        timestamp: new Date().toLocaleTimeString('en-US'),
        text:
          'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAH Hi I am flooding you with logs.',
      },
    });
  }

  await delay(400);

  // NOTE(jim): Rapid fire 40 logs over 4 devices and tests if it scrolls to bottom.
  for (let i = 0; i < 40; i++) {
    props.dispatch({
      type: 'PUSH_LOG_TO_DEVICE',
      log: {
        deviceId: getRandomInt(1, 3),
        status: 'success',
        timestamp: new Date().toLocaleTimeString('en-US'),
        text:
          'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAH Hi I am flooding you with logs.',
      },
    });
  }
};
